snowfall/models/transformer.py

```diff
@@ -0,0 +1,759 @@
+            target = target.masked_fill(ignore, 0)  # avoid -1 index
+            true_dist.scatter_(1, target.unsqueeze(1), self.confidence)
+        kl = self.criterion(torch.log_softmax(x, dim=1), true_dist)
+        denom = total if self.normalize_length else batch_size
+        return kl.masked_fill(ignore.unsqueeze(1), 0).sum() / denom
+
+
+def encoder_padding_mask(supervision_segments: Optional[Tensor] = None) -> Optional[Tensor]:
+    """Make mask tensor containing indices of padded part.
+
+    Args:
+        supervision_segments: Supervison in lhotse format, including 'sequence_idx',
+                            'start_frame', and 'num_frames'. 
+
+    Returns:
+        Tensor: Mask tensor of dimension (batch_size, input_length), True denote the masked indices.
+    """
+    if supervision_segments == None:
+        return None
+
+    indices = torch.argsort(supervision_segments[:, 1], descending=False)
+    supervision_segments = supervision_segments[indices]
+
+    indices = torch.argsort(supervision_segments[:, 0], descending=False)
+    supervision_segments = supervision_segments[indices]
+
+    previous_idx = -1
+    cur_length = 0
+    lengths = []
+    for sequence_idx, start_frame, num_frames in supervision_segments:
+        cur_length = start_frame + num_frames
+        if sequence_idx == previous_idx:
+            lengths[-1] = cur_length
+        else:
+            lengths.append(cur_length)
+        previous_idx = sequence_idx
+
+    bs = int(len(lengths))
+    maxlen = int(max(lengths))
+    seq_range = torch.arange(0, maxlen, dtype=torch.int64)
+    seq_range_expand = seq_range.unsqueeze(0).expand(bs, maxlen)
+    seq_length_expand = seq_range_expand.new(lengths).unsqueeze(-1)
+    mask = seq_range_expand >= seq_length_expand
+
+    return mask
+
+
+def decoder_padding_mask(ys_pad: Tensor, ignore_id: int = -1) -> Tensor:
+    """Generate a length mask for input. The masked position are filled with bool(True),
+        Unmasked positions are filled with bool(False).
+
+    Args:
+        ys_pad: padded tensor of dimension (batch_size, input_length).
+        ignore_id: the ignored number (the padding number) in ys_pad
+
+    Returns:
+        Tensor: a mask tensor of dimension (batch_size, input_length).
+    """
+    ys_mask = ys_pad == ignore_id
+    return ys_mask
+
+
+def get_normal_transcripts(supervision: Dict, words: k2.SymbolTable, oov: str = '<UNK>') -> List[List[int]]:
+    """Get normal transcripts (1 input recording has 1 transcript) from lhotse cut format.
+    Achieved by concatenate the transcripts corresponding to the same recording.
+
+    Args:
+        supervision : Supervison in lhotse format, i.e., batch['supervisions']
+        words: The word symbol table.
+        oov: Out of vocabulary word.
+
+    Returns:
+        List[List[int]]: List of concatenated transcripts, length is batch_size
+    """
+    
+    texts = [[token if token in words else oov
```

Comment:
... I wonder how the joint attention training would behave if the targets were phone, rather than character sequences?

Reply:
The lexicon has alternate pronunciations and optional silences- hard to encode into a single sequence.
I think this part is used just for an auxiliary task for regularization, we don't actually decode with it.
I would like to find an alternative way to regularize the model, as this seems kind of overkill for that purpose.

Reply:
@pzelasko  Currently the targets of joint attention training are also phones. Phone level targets were generated from word level targets using function `get_hierarchical_targets`. I haven't tried characters for targets yet.

@danpovey  One word sequence could lead to multiple phone sequences with this lexicon. But with function `get_hierarchical_targets`, it seems a deterministic phone sequence could be chosen. I think it won't affect much since we only use it for regularization. This auxiliary task is indeed too expensive to do. Maybe I should turn it off by default and replace it when we find an alternative way?

Reply:
Oh, cool, I missed that!